README.md
# Tephra
A tool for discovering transposable elements and describing patterns of genome evolution

Build|Coverage|Github Version|Docker Image Size|Docker Pulls
---|---|---|---|---
[![CI](https://github.com/sestaton/tephra/actions/workflows/main.yml/badge.svg)](https://github.com/sestaton/tephra/actions/workflows/main.yml) | [![codecov](https://codecov.io/gh/sestaton/tephra/branch/master/graph/badge.svg?token=MH622EAL1G)](https://codecov.io/gh/sestaton/tephra) | [![GitHub version](https://badge.fury.io/gh/sestaton%2Ftephra.svg)](https://badge.fury.io/gh/sestaton%2Ftephra) | ![Docker Image Size](https://badgen.net/docker/size/sestaton/tephra?icon=docker&label=image%20size) | ![Pulls](https://img.shields.io/docker/pulls/sestaton/tephra.svg)

### What is Tephra?

Tephra is a command line application to annotate [transposable elements](http://en.wikipedia.org/wiki/Transposable_element) from a genome assembly. The goal is to provide a high quality set of *de novo* annotations for all transposon types, describe the structure and evolution of those sequences, and do it without a reference set of transposon sequences (therefore being unbiased as possible).

**RECOMMENDED USAGE**

With [Docker](https://www.docker.com/), you can create a container to run Tephra with the following command:

    docker run -it --name tephra-con -v $(pwd)/db:/db:Z sestaton/tephra

That will create a container called `tephra-con` and start an interactive shell. The above assumes you have a directory called `db` in the working directory that contains your database files and the Tephra configuration. To run the full analysis, change to the mounted directory with `cd /db` in your container and run the following command:

    tephra all -c tephra_config.yml

I recommend using `nohup` and then logging out, which will allow you to leave the container running in the background.

If you cannot use Docker, please see the [INSTALL](https://github.com/sestaton/tephra/blob/master/INSTALL.md) file included with this distribution to install Tephra on various operating systems.

**BASIC USAGE**

Tephra is a command-line program only for now. The command `tephra` itself controls all the action of the subcommands, which perform specific tasks. Typing the command `tephra` will show the available commands. Here is an example,

    $ tephra 

    Tephra version 0.14.0
    
    Copyright (C) 2015-2023 S. Evan Staton
    LICENSE -- MIT

    Citation: Staton, SE. 2023. https://github.com/sestaton/tephra

    Name:
         Tephra - A tool for discovering transposable elements and describing
         patterns of genome evolution
    
    Description:
         This is an application to find transposable elements based on structural and sequence similarity features,
         group those elements into recognized (superfamilies) and novel (families) taxonomic groups,
         and infer patterns of evolution.
    
    -------------------------------------------------------------------------------------------
    USAGE: tephra <command> [options]
    
    Available commands:
         
                age: Calculate the age distribution of LTR or TIR transposons.
                all: Run all subcommands and generate annotations for all transposon types.
       classifyltrs: Classify LTR retrotransposons into superfamilies and families.
       classifytirs: Classify TIR transposons into superfamilies.
      findfragments: Search a masked genome with a repeat database to find fragmented elements.
      findhelitrons: Find Helitons in a genome assembly.
           findltrs: Find LTR retrotransposons in a genome assembly.
        findnonltrs: Find non-LTR retrotransposons in a genome assembly.  
           findtirs: Find TIR transposons in a genome assembly.
          findtrims: Find TRIM retrotransposons in a genome assembly.
          illrecomb: Characterize the distribution of illegitimate recombination in a genome.
            maskref: Mask a reference genome with transposons.
         reannotate: Transfer annotations from a reference set of repeats to Tephra annotations.
            sololtr: Find solo-LTRs in a genome assembly.
               info: Show version information for all external programs configured and used by Tephra.
	       
    Most common usage:
    
        tephra all -c tephra_config.yml
    
     That will produce a FASTA and GFF3 of all intact and fragmented transposons in the genome,
     and generate a table of annotation results.
    
    To get the configuration file, run:
    
        wget https://raw.githubusercontent.com/sestaton/tephra/master/config/tephra_config.yml
    
    To see information about a subcommand, run:
    
        tephra <command> --help
    
    To get more detailed information, run:
    
        tephra <command> --man


Typing a subcommand will show the usage of that command, for example:

    $ tephra findnonltrs

    [ERROR]: Required arguments not given.
    
    Name:
         tephra findnonltrs - Find non-LTRs retrotransposons in a genome assembly.
    
    Description:
         Find non-LTR retrotransposons in a reference genome, classify them into known superfamilies, 
         and generate a GFF file showing their locations and properties.
    
    USAGE: tephra findnonltrs [-h] [-m]
        -m --man      :   Get the manual entry for a command.
        -h --help     :   Print the command usage.
    
    Required:
        -g|genome     :   The genome sequences in FASTA format to search for non-LTR-RTs. 
        -o|gff        :   The GFF3 outfile to place the non-LTRs found in <genome>.
    
    Options:
        -r|reference  :   The non-masked reference genome for base correction.
        -d|outdir     :   The location to place the results.
        -p|pdir       :   Location of the HMM models (Default: configured automatically).
        -t|threads    :   The number of threads to use for BLAST searches (Default: 1).
        -v|verbose    :   Display progress for each chromosome (Default: no).


**SUPPORT AND DOCUMENTATION**

You can get usage information at the command line with the following command:

    perldoc tephra

The `tephra` program will also print a diagnostic help message when executed with no arguments, and display the available subcommands.

You can also look for information at:

    Tephra wiki
            https://github.com/sestaton/tephra/wiki

    Tephra issue tracker
            https://github.com/sestaton/tephra/issues

 
**CITATION**

A manuscript is in preparation, which includes a description of the all the methods and their uses, a comparison to other programs, and results from model systems. These will be provided in some form ahead of publication, as soon as they are available.

For now, please cite the github URL of this repo if you use Tephra. Thank you. 

**CURRENT STATUS**

Please check the [wiki](https://github.com/sestaton/tephra/wiki) for progress updates.

**CONTRIBUTING**

I welcome any comments, bug reports, feature requests, or contributions to the development of the project. Please submit a new issue (preferred) or send me an email and I would be happy to talk about Tephra or transposons.

**LICENSE AND COPYRIGHT**

Part of this project uses code from [MGEScan-nonLTR](http://darwin.informatics.indiana.edu/cgi-bin/evolution/nonltr/nonltr.pl), which is released under the GPL license. With permission of the authors, this code is packaged with Tephra. Below is the copyright for MGEScan-nonLTR:

    Copyright (C) 2015. See the LICENSE file for license rights and limitations (GPL v3).

    This program is part of MGEScan.

    MGEScan is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along with this program. If not, see http://www.gnu.org/licenses/.

The license for Tephra is below:

Copyright (C) 2015-2023 S. Evan Staton

This program is distributed under the MIT (X11) License, which should be distributed with the package.
If not, it can be found here: http://www.opensource.org/licenses/mit-license.php



Changes
Revision history for Tephra

Version	Date	Location
0.14.0	04/22/2023	Saskatoon, SK

New features:
 - Add 'info' subcommand to print a table of configuration info (Perl and Tephra versions, and versions of all installed
   external programs.
   
Bug fixes:
 - Adjust regex for getting divergence from PAML. This is something that unfortunately needs to be adjusted with
   updates to PAML.
 - Fixed the console messages when configuring and installing required packages so the output is much cleaner.
 
Misc:
 - Update Perl deps when building on v5.30+ (Pod::Find).
 - Add required modules for fetching muscle (and other packages) over https (Net::SSLeay and IO::Socket::SSL)
   during the initial configuration.
 - Change how the tephra command is being called when testing the full pipeline. Now it can be tested in place rather
   than expecting to be installed, which is much more desirable (avoids version conflicts and having to install the package
   to test the command).
 - Pin genometools version to v1.6 to ensure we are working with a stable version.
 - Update PAML from v4.8 to v4.10.6.
 - Capture the noisy output from EMBOSS compile process. Same with HTSlib and Tephra translate program.
 - Pin muscle to v3.8.31 and compile from source because the available binaries have issues in my tests.
 - Remove use of Travis-CI and switch to Github Actions for CI workflow.
 - Add coverage report from Codecov.
 - Add test file for new 'info' subcommand to evaluate results that are returned.
 - Update Docker and Github Actions to use cpm, which is much faster than to build than cpanm.
 
0.13.2	07/20/2020	Vancouver, BC

New features:
 - Add correct Sequence Ontology terms for TRIM and LARD elements in GFF3 output. I requested these terms previously
   and the changes in this version bring the GFF3 output of Tephra to be in agreement with the latest specifications.

- Misc:
 - Update singularity usage method to build locally and avoid permissions issues when pulling pre-build images.
0.13.1	07/03/2020	Vancouver, BC

Bug fixes:
 - Fix issue with PATHs to Tephra configured programs (vmatch and mkvtree) not being set in *Annotation::MakeExemplars class.

Misc:
 - Updated TODO for this issue because it is hard to test (more details there).
 
0.13.0	06/30/2020	Vancouver, BC

New features:
 - Refactor build system to not used shared libraries and only included required dependency files.

Misc:
 - Correct alignment of reporting in log for LTR elements. 

0.12.6	04/07/2020	Vancouver, BC

New features:
 - Add parallel search of both strands for non-LTR retrotransposons, which reduces
   user runtime by about 50% if multiple CPUs are available.
 - Use multiple CPUs (if requested) for HMM model searches for non-LTR elements.
 - Greatly simplify call to class methods for finding non-LTRs based on the use of 
   new class attributes (req. fewer method arguments).
 - Add resusable for writing/annotating elements (both for families and singleton
   elements) in Tephra::Classify::Any.
 - Warn now when filtering Helitron/non-LTR elements based on N%.
 - Add --debug option for monitoring config/install process.

Bug fixes:
 - Adjust non-LTR element N% filtering logic to keep FASTA and GFF3 IDs properly in sync.
 - Improve logic for classifying Helitron and non-LTR retrotransposon families by storing
   full sequence IDs so no regex or transformations are needed.
 - Adjust method for writing elements in Tephra::Classify::Any to include all elements in
   a cluster (query and hits) above the threshold.
 - As with other family-level classification methods in Tephra, search for families only within
   a superfamily for Helitrons and non-LTR elements.

Misc:
- Add role method (*store_seq()) for hashing sequences+IDs that can be used in other classes.

0.12.5	08/26/2019	Vancouver, BC

New features:
 - Add method for filtering LTR/TIR/TRIM elements against a user-provided gene set to
   remove spurious predictions that are likely tandemly repeated genes. 
 - Handle compressed input for 'findltrs', 'findtirs', 'findtrims', and 'findfragments'
   commands.

Bug fixes:
 - Configure all external programs during installation instead of relying on those in PATH.
   This fixes the issue of not being able to find programs when running in a container.
   Fixes #41 on github.
 - Correct the number of elements discovered under different relaxed and strict conditions.
   What was being reported was the filtered numbers (correct for the total, but not the correct
   number initially identified).
 - Remove use of FTP protocol for configuring dependencies (and drop Net::FTP requirement). This
   protocol is not supported in all build environments, such as Travis, so it is better to use
   wget as a permanent replacement than to keep debugging FTP connection issues.
 - Configure install of EMBOSS during testing instead using package manager. Also, remove package
   installs of BLAST+ and EMBOSS to speed things up since we configure them locally.

0.12.4	05/27/2019	Vancouver, BC

New features:
 - Add protein domain coordinates for non-LTR elements to GFF output.
 - Parallelize search for EN and RVT domains in 'findnonltrs' command.

Bug fixes:
- Fix method for how the number of non-LTR and Helitron families are calculated. Previously,
  there may have erroneously been single-copy families and singletons that should have been
  in families due to how the superfamilies where being indexed and combined.
- In the *Annotation::Util class, there was a bug in getting codes for non-LTR superfamilies
  with some codes being redefined.

0.12.3	01/04/2019	Vancouver, BC

New features:
- Add support for compressed input to 'all' command (and add tests with compressed data).
- Add new suite of dev tests for the 'findtirs' and 'classifytirs' commands to
  ensure the bugs reported below with respect the element/family numbers are
  correct.
- New methods for naming MITEs and LARDs so they are numbered sequentially instead

lib/Tephra/Config.pm
package Tephra::Config;

use 5.010;
use Moose;
use MooseX::Types::Path::Class;
use Cwd;
use Config;
use File::Spec;
use File::Find;
use File::Copy qw(copy move);
use File::Path qw(make_path remove_tree);
use File::Basename;
use Path::Class::File;
use HTML::TreeBuilder;
use HTTP::Tiny;
use Net::FTP;
use Log::Any qw($log);
use namespace::autoclean;

=head1 NAME

Tephra::Config - Class for setting up Tephra dependencies

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';

has basedir => (
    is       => 'ro',
    isa      => 'Path::Class::Dir',
    required => 0,
    coerce   => 1,
    default  => sub {
	return Path::Class::Dir->new($ENV{HOME})
    },
);

has workingdir => ( 
    is       => 'ro', 
    isa      => 'Path::Class::Dir', 
    required => 1, 
    coerce   => 1 
);

sub configure_root {
    my $self = shift;
    my $root = $self->basedir;
    
    unless (-e $root) {
	make_path($root, {verbose => 0, mode => 0711,});
    }
    
    # we don't want to reconfigure every time the tests run
    my $gt      = File::Spec->catfile($root, 'gt', 'bin', 'gt');
    my $hscan   = File::Spec->catfile($root, 'helitronscanner', 'HelitronScanner', 'HelitronScanner.jar');
    my $hmmbin  = File::Spec->catdir($root, 'hmmer-2.3.2', 'bin');
    my $moddir  = File::Spec->catdir($root, 'pHMM');
    my $chrdir  = File::Spec->catdir($root, 'hmm');
    my $mgescan = File::Spec->catfile($chrdir, 'tephra-MGEScan');
    my $transla = File::Spec->catfile($chrdir, 'tephra-translate');
    my $clw     = File::Spec->catfile($root, 'clustalw-2.1', 'bin', 'clustalw2');
    my $pamlbin = File::Spec->catdir($root, 'paml4.8', 'bin');
    my $transeq = File::Spec->catdir($root, 'EMBOSS-6.5.7', 'bin');
    my $samtool = File::Spec->catfile($root, 'samtools-1.2', 'samtools');
    my $blastph = File::Spec->catdir($root, 'ncbi-blast-2.2.31+', 'bin');

    unless (-e $gt) {
	$gt = $self->fetch_gt_exes;
    }
    
    unless (-e $hscan) {
	$hscan = $self->fetch_hscan;
    }
    
    unless (-e $hmmbin) {
	$hmmbin = $self->fetch_hmmer2;
    }
    
    unless (-e $moddir) {
	$moddir = $self->fetch_hmm_models;
    }
    
    unless (-e $chrdir) {
	$chrdir = $self->make_chrom_dir;
    }
    
    unless (-e $mgescan && -e $transla) {
	($mgescan, $transla) = $self->build_mgescan;
    }
    
    unless (-e $clw) {
	$clw = $self->fetch_clustalw2;
    }
    
    unless (-e $pamlbin) {
	$pamlbin = $self->fetch_paml;
    }
    
    unless (-e $transeq) {
	$transeq = $self->fetch_paml;
    }

    unless (-e $samtool) {
        $samtool = $self->fetch_samtools;
    }

    unless (-e $blastph) {
        $blastph = $self->fetch_blast;
    }

    return ({
	gt        => $gt, 
	hscandir  => $hscan, 
	hmmerbin  => $hmmbin, 
	modeldir  => $moddir, 
	hmmdir    => $chrdir, 
	mgescan   => $mgescan, 
	transcmd  => $transla, 
	clustalw  => $clw, 
	pamlbin   => $pamlbin,
	transeq   => $transeq,
	samtools  => $samtool,
	blastpath => $blastph });

}

sub fetch_gt_exes {
    my $self = shift;
    my $root = $self->basedir;
    my $wd   = $self->workingdir;
    
    my $host = 'http://genometools.org';
    my $dir  = 'pub/binary_distributions';
    my $file = 'gt_distlisting.html';
    $self->fetch_file($file, $host."/".$dir);
    
    my $tree = HTML::TreeBuilder->new;
    $tree->parse_file($file);
    
    my ($dist, $ldist, $ldir);
    for my $tag ($tree->look_down(_tag => 'a')) {
	if ($tag->attr('href')) {
	    if ($tag->as_text =~ /Linux_x86_64-64bit-barebone.tar.gz\z/) {
		$dist = $tag->as_text;
		my $archive = join "/", $host, $dir, $dist;
		$self->fetch_file($dist, $archive);
		
		$ldist = $dist;
		$ldist =~ s/\.tar.gz\z//;
		$ldir = File::Spec->catdir($root, 'gt');
		
		system("tar xzf $dist") == 0 or die $!;
		
		move $ldist, $ldir or die "Move failed: $!";
		unlink $dist;
	    }
	}
    }
    unlink $file;
    
    my $gt = File::Spec->catfile($ldir, 'bin', 'gt');

    return $gt
}

sub fetch_hscan {
    my $self = shift;
    my $root = $self->basedir;
    my $wd   = $self->workingdir;
    
    my $host = 'http://sourceforge.net';
    my $dir  = 'projects/helitronscanner/files/HelitronScanner_V1.0.zip/download';
    my $ldir = File::Spec->catdir($root, 'helitronscanner');
    make_path( $ldir, {verbose => 0, mode => 0771,} );
    my $file = 'HelitronScanner.zip';
    my $path = File::Spec->catfile($ldir, $file);
    $self->fetch_file($path, $host."/".$dir);
    chdir $ldir or die $!;
    system("unzip $file 2>&1 > /dev/null") == 0 or die $!;
    
    my $cwd   = getcwd();
    my $hscan = File::Spec->catfile($cwd, 'HelitronScanner', 'HelitronScanner.jar');
    chdir $wd;
    
     return $hscan;
}

sub fetch_samtools {
    my $self = shift;
    my $root = $self->basedir;
    my $wd   = $self->workingdir;

    my $host = 'http://sourceforge.net';
    my $dir  = 'projects/samtools/files/samtools/1.2/samtools-1.2.tar.bz2/download';
    my $file = 'samtools-1.2.tar.bz2';
    my $path = File::Spec->catfile($root, $file);
    chdir $root or die $!;

    $self->fetch_file($path, $host."/".$dir);
    system("tar xjf $file 2>&1 > /dev/null") == 0 or die $!;
    unlink $file if -e $file;

    my $dist = 'samtools-1.2';
    chdir $dist or die $!;
    system("make -j4 2>&1 > /dev/null") == 0 or die $!;

    my $cwd = getcwd();
    my $samtools = File::Spec->catfile($cwd, 'samtools');
    chdir $wd or die $!;

    return $samtools;
}

sub fetch_blast {
    my $self = shift;
    my $root = $self->basedir;
    my $wd   = $self->workingdir;

    chdir $root or die $!;
    my $host = 'ftp.ncbi.nlm.nih.gov';
    my $ftp = Net::FTP->new($host, Passive => 1, Debug => 0)
	or die "Cannot connect to $host: $@";

    $ftp->login or die "Cannot login ", $ftp->message;

    my $dir  = '/blast/executables/blast+/LATEST';
    $ftp->cwd($dir)
	or die "Cannot change working directory ", $ftp->message;

    my $file = 'ncbi-blast-2.2.31+-x64-linux.tar.gz';

    $ftp->binary();
    my $rsize = $ftp->size($file) or die "Could not get size ", $ftp->message;
    $ftp->get($file) or die "get failed ", $ftp->message;
    my $lsize = -s $file;
    die "Failed to fetch complete file: $file (local size: $lsize, remote size: $rsize)"
	unless $rsize == $lsize;

    my $ldir = 'ncbi-blast-2.2.31+';
    system("tar xzf $file 2>&1 > /dev/null") == 0 or die $!;
    unlink $file if -e $file;
    chdir $ldir or die $!;

    my $cwd = getcwd();
    my $blastpath = File::Spec->catfile($cwd, 'bin');

    return $blastpath;
}

sub fetch_hmmer2 {
    my $self = shift;
    my $root = $self->basedir;
    my $wd   = $self->workingdir;
    
    my $urlbase = 'http://selab.janelia.org'; 
    my $dir     = 'software';
    my $tool    = 'hmmer';
    my $version = '2.3.2';
    my $file    = 'hmmer-2.3.2.tar.gz';
    my $url     = join "/", $urlbase, $dir, $tool, $version, $file;
    my $outfile = File::Spec->catfile($root, $file);
    $self->fetch_file($outfile, $url);

    chdir $root;
    my $dist = 'hmmer-2.3.2';
    system("tar xzf $file") == 0 or die "tar failed: $!";
    chdir $dist;
    my $cwd = getcwd();
    system("./configure --enable-threads --prefix=$cwd 2>&1 > /dev/null") == 0
	or die "configure failed: $!";
    system("make -j4 2>&1 >/dev/null") == 0 
	or die "make failed: $!";
    system("make install 2>&1 >/dev/null") == 0
	 or die "make failed: $!";
    my $hmmbin = File::Spec->catdir($cwd, 'bin');
    my $distfile = File::Spec->catfile($root, $file);
    unlink $distfile;
    chdir $wd;
    
    return $hmmbin;
}

sub fetch_clustalw2 {
    my $self = shift;
    my $root = $self->basedir;
    my $wd   = $self->workingdir;
    
    my $urlbase = 'http://www.clustal.org';
    my $dir     = 'download';
    my $tool    = 'current';
    my $file    = 'clustalw-2.1.tar.gz';
    my $url     = join "/", $urlbase, $dir, $tool, $file;
    my $outfile = File::Spec->catfile($root, $file);
    $self->fetch_file($outfile, $url);

    chdir $root;
    my $dist = 'clustalw-2.1';
    system("tar xzf $file") == 0 or die "tar failed: $!";
    chdir $dist;
    my $cwd = getcwd();
    system("./configure --prefix=$cwd 2>&1 > /dev/null") == 0
	or die "configure failed: $!";
    system("make -j4 2>&1 > /dev/null") == 0 
	or die "make failed: $!";
    system("make install 2>&1 > /dev/null") == 0
	or die "make failed: $!";
    
    my $clw = File::Spec->catdir($cwd, 'bin', 'clustalw2');
    my $distfile = File::Spec->catfile($root, $file);
    unlink $distfile;
    chdir $wd;

    return $clw;
}

sub fetch_paml {
    my $self = shift;
    my $root = $self->basedir;
    my $wd   = $self->workingdir;

    my $urlbase = 'http://abacus.gene.ucl.ac.uk';
    my $dir     = 'software';
    my $file    = 'pamlX1.3.1+paml4.8a-win32.tgz';
    my $url     = join "/", $urlbase, $dir, $file;
    my $outfile = File::Spec->catfile($root, $file);
    $self->fetch_file($outfile, $url);

    chdir $root;
    my $dist  = 'paml4.8';
    my $xdist = 'pamlX';
    system("tar xzf $file") == 0 or die "tar failed: $!";
    remove_tree( $xdist, { safe => 1 } );
    unlink $file;

    chdir $dist;
    my $cwd = getcwd();
    my $bin = File::Spec->catdir($cwd, 'bin');
    my @exes;
    find( sub { push @exes, $File::Find::name if -f and /\.exe$/ }, $bin );
    unlink @exes;
    chdir 'src';
    system("make -j4 2>&1 >/dev/null") == 0 
	or die "make failed: $!";

    my @exelist = ('yn00', 'baseml', 'basemlg', 'mcmctree', 'pamp', 'evolver', 'infinitesites', 'codeml');
    my $rootdir = File::Spec->catdir($root, 'paml4.8', 'bin');
    unless ( -d $rootdir ) {
	make_path( $rootdir, {verbose => 0, mode => 0771,} );
    }

    for my $file (@exelist) {
	copy $file, $rootdir or die "Copy failed: $!";
    }

    my @nonexes = map { File::Spec->catfile($rootdir, $_) } @exelist;
    my $cnt = chmod 0755, @nonexes;
    
    if ($cnt == @exelist) {
	return $rootdir;
    }
}

sub fetch_emboss {
    my $self = shift;
    my $root = $self->basedir;
    my $wd   = $self->workingdir;

    # this is to avoid building each time
    my @path = split /:|;/, $ENV{PATH};    
    for my $p (@path) {
	my $transeq  = File::Spec->catfile($p, 'transeq');
	if (-e $transeq && -x $transeq) {
	    return $transeq;
	}
    }

    my $urlbase = 'ftp://emboss.open-bio.org';
    my $dir     = 'pub';
    my $tool    = 'EMBOSS';
    my $release = 'old';
    my $version = '6.5.0';
    my $file    = 'EMBOSS-6.5.7.tar.gz';
    my $url     = join "/", $urlbase, $dir, $tool, $release, $version, $file;
    my $outfile = File::Spec->catfile($root, $file);
    $self->fetch_file($outfile, $url);

    chdir $root;
    my $dist = 'EMBOSS-6.5.7';
    system("tar xzf $file") == 0 or die "tar failed: $!";
    chdir $dist;
    my $cwd = getcwd();
    system("./configure --prefix=$cwd 2>&1 > /dev/null") == 0
	or die "configure failed: $!";
    system("make -j4 2>&1 > /dev/null") == 0 
	or die "make failed: $!";
    system("make install 2>&1 > /dev/null") == 0
	or die "make failed: $!";
    
    my $transeq = File::Spec->catdir($cwd, 'bin', 'transeq');
    my $distfile = File::Spec->catfile($root, $file);
    unlink $distfile;
    chdir $wd;

    return $transeq;
}

sub fetch_hmm_models {
    my $self = shift;
    my $root = $self->basedir;
    my $wd   = $self->workingdir;
   
    chdir $wd;
    my $file = 'pHMM.tar.gz';
    my $dist = File::Spec->catfile('build', $file);
    copy $dist, $root or die "Copy failed: $!";
    chdir $root;
    system("tar xzf $file") == 0 or die $!;
    unlink $file;

    my $dir = File::Spec->catfile($root, 'pHMM');

    return $dir;
}

sub make_chrom_dir {
    my $self = shift;
    my $root = $self->basedir;

    my $hmm_dir = File::Spec->catdir($root, 'hmm');
    unless ( -d $hmm_dir ) {
        make_path( $hmm_dir, {verbose => 0, mode => 0771,} );
    }

    my $chr_file = File::Spec->catfile($hmm_dir, 'chr.hmm');
    open my $out, '>', $chr_file;
    say $out "Symbol= 4";
    say $out "State= 33";
    say $out "Transition= 73";
    say $out join "\t", '0', '1', '0.0455';
    say $out join "\t", '0', '3', '0.0455';
    say $out join "\t", '0', '4', '0.0455';
    say $out join "\t", '0', '6', '0.0455';
    say $out join "\t", '0', '7', '0.0455';
    say $out join "\t", '0', '9', '0.0455';
    say $out join "\t", '0', '10', '0.0455';
    say $out join "\t", '0', '12', '0.0455';
    say $out join "\t", '0', '13', '0.0455';
    say $out join "\t", '0', '15', '0.0455';
    say $out join "\t", '0', '16', '0.0455';
    say $out join "\t", '0', '18', '0.0455';
    say $out join "\t", '0', '19', '0.0455';
    say $out join "\t", '0', '21', '0.0455';
    say $out join "\t", '0', '22', '0.0455';
    say $out join "\t", '0', '24', '0.0455';
    say $out join "\t", '0', '25', '0.0455';
    say $out join "\t", '0', '27', '0.0455';
    say $out join "\t", '0', '28', '0.0455';
    say $out join "\t", '0', '30', '0.0455';
    say $out join "\t", '0', '31', '0.0455';
    say $out join "\t", '0', '32', '0.0455';
    say $out join "\t", '1', '0', '0.5';
    say $out join "\t", '1', '2', '0.5';
    say $out join "\t", '2', '3', '1.0';
    say $out join "\t", '3', '0', '1.0';
    say $out join "\t", '4', '0', '0.5';
    say $out join "\t", '4', '5', '0.5';
    say $out join "\t", '5', '6', '1.0';
    say $out join "\t", '6', '0', '1.0';
    say $out join "\t", '7', '0', '0.5';
    say $out join "\t", '7', '8', '0.5';
    say $out join "\t", '8', '9', '1.0';
    say $out join "\t", '9', '0', '1.0';
    say $out join "\t", '10', '0', '0.5';
    say $out join "\t", '10', '11', '0.5';
    say $out join "\t", '11', '12', '1.0';
    say $out join "\t", '12', '0', '1.0';
    say $out join "\t", '13', '0', '0.5';
    say $out join "\t", '13', '14', '0.5';
    say $out join "\t", '14', '15', '1.0';
    say $out join "\t", '15', '0', '1.0';
    say $out join "\t", '16', '0', '0.5';
    say $out join "\t", '16', '17', '0.5';
    say $out join "\t", '17', '18', '1.0';
    say $out join "\t", '18', '0', '1.0';
    say $out join "\t", '19', '0', '0.5';
    say $out join "\t", '19', '20', '0.5';
    say $out join "\t", '20', '21', '1.0';
    say $out join "\t", '21', '0', '1.0';
    say $out join "\t", '22', '0', '0.5';
    say $out join "\t", '22', '23', '0.5';
    say $out join "\t", '23', '24', '1.0';
    say $out join "\t", '24', '0', '1.0';
    say $out join "\t", '25', '0', '0.5';
    say $out join "\t", '25', '26', '0.5';
    say $out join "\t", '26', '27', '1.0';
    say $out join "\t", '27', '0', '1.0';
    say $out join "\t", '28', '0', '0.5';
    say $out join "\t", '28', '29', '0.5';
    say $out join "\t", '29', '30', '1.0';
    say $out join "\t", '30', '0', '1.0';
    say $out join "\t", '31', '0', '1.0';
    say $out join "\t", '32', '0', '1.0';
    say $out "Pi= 33";
    say $out "0.1";
    print $out "0.05\n" x 31;
    print $out '0.05';
    close $out;

    return $hmm_dir;
}

sub build_mgescan {
    my $self = shift;
    my $root = $self->basedir;
    my $wd   = $self->workingdir;

    my $hmm_dir = File::Spec->catdir($root, 'hmm');
    my $src_dir = File::Spec->catdir($wd, 'src');
    my $mgexe   = 'tephra-MGEScan';
    my $trexe   = 'tephra-translate';
    my $mgescan = File::Spec->catfile($hmm_dir, $mgexe);
    my $transla = File::Spec->catfile($hmm_dir, $trexe);

    chdir $src_dir;
    system("make clean -f mgescan-makefile 2>&1 >/dev/null") == 0 
	or die "make failed: $!";
    system("make -f mgescan-makefile 2>&1 >/dev/null") == 0 
	or die "make failed: $!";
    system("make clean -f translate-makefile 2>&1 >/dev/null") == 0
        or die "make failed: $!";
    system("make all -f translate-makefile 2>&1 >/dev/null") == 0
        or die "make failed: $!";
    
    copy $mgexe, $hmm_dir or die "Copy failed: $!";
    copy $trexe, $hmm_dir or die "Copy failed: $!";
    my $cnt = chmod 0755, $mgescan, $transla;
    
    if ($cnt == 2 && -e $mgescan && -e $transla) {
	return ($mgescan, $transla);
    }
}

sub fetch_file {
    my $self = shift;
    my ($file, $endpoint) = @_;
    unless (-e $file) {
	my $response = HTTP::Tiny->new->get($endpoint);
	unless ($response->{success}) {
	    die "Can't get url $endpoint -- Status: ", $response->{status}, 
	        " -- Reason: ", $response->{reason};
	}
	open my $out, '>', $file;
	print $out $response->{content};
	#sleep 1;
	close $out;
    }
}

=head1 AUTHOR

S. Evan Staton, C<< <statonse at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests through the project site at 
L<https://github.com/sestaton/tephra/issues>. I will be notified,
and there will be a record of the issue. Alternatively, I can also be 
reached at the email address listed above to resolve any questions.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Tephra::Config


=head1 LICENSE AND COPYRIGHT

Copyright (C) 2015- S. Evan Staton

This program is distributed under the MIT (X11) License, which should be distributed with the package. 
If not, it can be found here: L<http://www.opensource.org/licenses/mit-license.php>

=cut 

__PACKAGE__->meta->make_immutable;

1;
